internal/controller/imageprefetch_controller.go

```diff
@@ -2,45 +2,508 @@ package controller
+			Status:  metav1.ConditionTrue,
+			Reason:  "ImagePrefetchProgressing",
+			Message: "Waiting for all nodes to pull the image",
+		})
+	}
+
+	return result, r.Status().Update(ctx, imgPrefetch)
+}
+
+type NodeImageSetStatus struct {
+	desiredNodes    int
+	availableNodes  int
+	pullingNodes    int
+	pullFailedNodes int
+}
+
+func calculateStatus(selectNodes []string, nodeImageSets *ofenv1.NodeImageSetList) NodeImageSetStatus {
+	status := NodeImageSetStatus{}
+	status.desiredNodes = len(selectNodes)
+
+	for _, nodeImageSet := range nodeImageSets.Items {
+		if meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageAvailable) {
+			status.availableNodes++
+		}
+		if meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageDownloadFailed) {
+			status.pullFailedNodes++
+		}
+		if meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageDownloadComplete) &&
+			!meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageAvailable) {
+			status.pullingNodes++
+		}
+	}
+
+	return status
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ImagePrefetchReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	nodeImageSetHandler := handler.EnqueueRequestsFromMapFunc(
+		func(ctx context.Context, obj client.Object) []ctrl.Request {
+			nodeImageSet := obj.(*ofenv1.NodeImageSet)
+			imagePrefetch := &ofenv1.ImagePrefetch{}
+			err := r.Get(ctx, types.NamespacedName{
+				Namespace: nodeImageSet.Labels[constants.OwnerImagePrefetchNamespace],
+				Name:      nodeImageSet.Labels[constants.OwnerImagePrefetchName],
+			}, imagePrefetch)
+			if err != nil {
+				return nil
+			}
+
+			return []ctrl.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Namespace: imagePrefetch.Namespace,
+						Name:      imagePrefetch.Name,
+					},
+				},
+			}
+		})
+
+	nodeHandler := handler.EnqueueRequestsFromMapFunc(
+		func(ctx context.Context, obj client.Object) []ctrl.Request {
+			node := obj.(*corev1.Node)
+			imagePrefetchList := &ofenv1.ImagePrefetchList{}
+			err := r.List(ctx, imagePrefetchList)
+			if err != nil {
+				return nil
+			}
+
+			var requests []ctrl.Request
+			for _, imgPrefetch := range imagePrefetchList.Items {
+				if imgPrefetch.Spec.NodeSelector == nil {
+					continue
+				}
+
+				for key, value := range imgPrefetch.Spec.NodeSelector {
+					if node.Labels[key] == value {
+						requests = append(requests, ctrl.Request{
+							NamespacedName: types.NamespacedName{
+								Namespace: imgPrefetch.Namespace,
+								Name:      imgPrefetch.Name,
+							},
+						})
+					}
+				}
+			}
+
+			return requests
+		})
+
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&ofenv1.ImagePrefetch{}).
+		Watches(
+			&ofenv1.NodeImageSet{},
+			nodeImageSetHandler,
+			builder.WithPredicates(
+				predicate.Funcs{
+					CreateFunc: func(e event.CreateEvent) bool {
+						return true
+					},
```

Comment:
Is it necessary to watch the creation of NodeImageSet?

As far as I understand, the ImagePrefetch Controller need not do anything immediately after the creation of NodeImageSet.

Reply:
I modified it to not watch for create events since there's no process that needs to be executed immediately after creating a NodeImageSet.
https://github.com/cybozu-go/ofen/pull/8/commits/09249c5ddb268b8f7f0168954a8b3941cddb3161